Add vSphere remotemachineset actuator

twiest opened the pull request:
- [x] Add vSphere remotemachineset actuator
- [x] Add unit tests
- [x] Manually test against real vmware cluster

This PR is dependent on PR https://github.com/openshift/hive/pull/975 and must merge after.

Fixes: #714

staebler commented on Dockerfile:
```diff
@@ -31,6 +31,18 @@ RUN mkdir -p /home/hive && \
     chgrp -R 0 /home/hive && \
     chmod -R g=u /home/hive
 
+# This is so that we can write source certificate anchors during container start up.
```
Please add these to Dockerfile.dev as well.

twiest replied:
Done.

staebler commented on pkg/apis/hive/v1/validating-webhooks/machinepool_validating_admission_hook.go:
```diff
@@ -284,6 +285,11 @@ func validateMachinePoolSpecInvariants(spec *hivev1.MachinePoolSpec, fldPath *fi
+	if p := spec.Platform.VSphere; p != nil {
+		platforms = append(platforms, "vsphere")
+		allErrs = append(allErrs, validateVSphereMachinePoolPlatformInvariants(p, platformPath.Child("vsphere"))...)
+		numberOfMachineSets = 1 // no zone support on vSphere
```
I would prefer `numberOfMachineSets` to be 0 so that the error when `spec.Autoscaling.MinReplicas` is less than 1 is
```
minimum replicas must at least 1
```
instead of
```
minimum replicas must be at least the number of zones
```
particularly since vSphere does not have zones.

twiest replied:
Should this be changed for OpenStack as well?

@dgoodwin thoughts as well?

I'll change both if you guys agree.

staebler replied:
Yes, it should be changed for OpenStack as well.

staebler commented on pkg/clusterresource/builder.go:
```diff
@@ -158,9 +155,7 @@ func (o *Builder) Build() ([]runtime.Object, error) {
-	if !o.SkipMachinePoolGeneration {
-		allObjects = append(allObjects, o.generateMachinePool())
-	}
+	allObjects = append(allObjects, o.generateMachinePool())
```
Let's put a nil check on the generated machine pool so that we don't have to add back the `SkipMachinePoolGeneration` when another platform comes along for which we don't yet have machine pool support.

twiest replied:
I'm not sure what you man here. `o.generatedMachinePool` doesn't seem like it will ever return a nil.

Can you explain your thoughts further here please?

staebler replied:
For the platforms that we have now, `generateMachinePool` does not return a nil. If we add support for a new platform for which Hive does not support MachinePools yet, then we need a way to skip machine pool generation. This should not come from a flag on the options: It should come from the builder implementation being used.

I am fine with leaving it as is for this PR and dealing with it when we do add support for the next platform.

twiest replied:
Ok, sounds good. We'll look at this in a future PR.

staebler commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -1608,6 +1608,8 @@ func (r *ReconcileClusterDeployment) mergePullSecrets(cd *hivev1.ClusterDeployme
 		if err != nil {
 			if !apierrors.IsNotFound(err) {
 				return "", err
+			} else {
```
Get rid of the check for whether it is a NotFound error.

staebler commented on pkg/controller/remotemachineset/vsphereactuator.go:
```diff
@@ -0,0 +1,74 @@
+	"fmt"
+
+	machineapi "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
```
Put "github.com/openshift/hive" imports in their own section at the end of the imports.

staebler commented on pkg/controller/remotemachineset/vsphereactuator.go:
```diff
@@ -0,0 +1,74 @@
+import (
+	"fmt"
+
+	machineapi "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
```
Combine this import with the other "github.com/openshift" imports that are not "github.com/openshift/hive".

staebler commented on pkg/controller/remotemachineset/vsphereactuator.go:
```diff
@@ -0,0 +1,74 @@
+
+	ic := &installertypes.InstallConfig{
+		Platform: installertypes.Platform{
+			VSphere: &installertypesvsphere.Platform{},
```
This is missing a lot of fields that are used by the installer to create the MachineSet. For example, it is missing the vCenter. Are those fields somehow obtained in-cluster and not needed?

https://github.com/openshift/installer/blob/8da3ffa23277b55b3ddbf53761063d22a30ce8a2/pkg/asset/machines/vsphere/machines.go#L66.

twiest replied:
For this, I copied the pattern used by the [OpenStack actuator](https://github.com/openshift/hive/blob/master/pkg/controller/remotemachineset/openstackactuator.go#L66-L75), which is basically only fill out the bare minimum used to be able to generate the machinesets.

When I originally looked through the `remotemachineset` controller code, I didn't see any place these values were used. But, that was because I was focusing on the remotemachineset Update code.

During a remotemachineset create call, this information is used, so I think you're correct that we should be filling this out.

I'm running some tests now to confirm.

@dgoodwin do you think that this is a problem with remotemachineset create in the OpenStack actuator as well?

twiest replied:
For posterity, the decision here was to fill out the install config with enough information so that the created machinesets had the same critical data that the machinesets created during install had. This was manually verified by comparing the data structures.

twiest commented:
Blocking until PR #975 is merged.

/block

twiest commented:
/hold

Blocking until PR #975 is merged.

twiest commented:
/retest

twiest commented:
/retest

twiest commented:
@staebler I believe I've addressed your feedback. PTAL.

twiest commented:
/retest

dgoodwin commented on pkg/clusterresource/vsphere.go:
```diff
@@ -111,7 +111,14 @@ func (p *VSphereCloudBuilder) addClusterDeploymentPlatform(o *Builder, cd *hivev
+	mp.Spec.Platform.VSphere = &hivev1vsphere.MachinePool{
+		NumCPUs:           2,
+		NumCoresPerSocket: 1,
+		MemoryMiB:         8192,
```
Seems a little low, but is this an installer default?

twiest replied:
I got those values from the default machinesets created by the installer.

This seems to be where the default value is stored, but I'm not 100% sure on that:
https://github.com/openshift/installer/blob/83213bd69d8a3e3937e4cdd7f7d1e6b734a535bf/upi/vsphere/variables.tf#L159-L162

staebler replied:
Those are the defaults that I see, too.

https://github.com/openshift/installer/blob/a6cd8414b02ba8cac400b8d6628e11900ca31687/pkg/asset/machines/worker.go#L130-L139

twiest commented:
@dgoodwin @staebler I've rebased and I believe I've addressed all of your feedback. PTAL.

/hold cancel

staebler commented on pkg/controller/remotemachineset/vsphereactuator.go:
```diff
@@ -10,7 +10,8 @@ import (
 	installertypesvsphere "github.com/openshift/installer/pkg/types/vsphere"
-	"github.com/openshift/machine-api-operator/pkg/apis/vsphereprovider/v1alpha1"
+
+	//"github.com/openshift/machine-api-operator/pkg/apis/vsphereprovider/v1beta1"
```
Remove this line.

staebler commented on pkg/controller/remotemachineset/vsphereactuator.go:
```diff
@@ -82,8 +83,5 @@ func (a *VSphereActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, pool
 		return nil, false, errors.Wrap(err, "failed to generate machinesets")
 	}
 
-	providerSpec := installerMachineSets[0].Spec.Template.Spec.ProviderSpec.Value.Object.(*v1alpha1.VSphereMachineProviderSpec)
```
Can you move this change to the other commit?

twiest commented:
@staebler Thanks for the feedback. I believe I've addressed the feedback. I also manually tested this against vsphere and the functionality worked as expected. PTAL.
